File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public void generate(int idx, int[] nums, int target, ArrayList<Integer> vec, List<List<Integer>> res) {
        if (target < 0) {
            return;
        }
        if (target == 0) {
            res.add(new ArrayList<>(vec));
            return;
        }
        if (idx == nums.length) {
            return;
        }
        
        vec.add(nums[idx]);
        generate(idx, nums, target - nums[idx], vec, res);
        vec.remove(vec.size() - 1);
        generate(idx + 1, nums, target, vec, res);
    }
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        generate(0, candidates, target, new ArrayList<>(), res);
        return res;
    }
}

File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void f(int i, vector<int>& nums, vector<int>& ds, vector<vector<int>>& res, int& n, int target) {
        if (i == n) {
            if (target == 0) {
                res.push_back(ds);
            }
            return;
        }
        
        f(i + 1, nums, ds, res, n, target);
        
        if (nums[i] <= target) {
            ds.push_back(nums[i]);
            f(i, nums, ds, res, n, target - nums[i]);
            ds.pop_back();
        }
    }
    
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        int n = nums.size();
        
        vector<vector<int>> res;
        
        vector<int> ds;
        
        f(0, nums, ds, res, n, target);
        
        return res;
    }
};